coral-common/src/main/java/com/linkedin/coral/common/functions/FunctionReturnTypes.java

```diff
@@ -28,13 +28,16 @@ private FunctionReturnTypes() {
 
   }
 
-  public static final SqlReturnTypeInference ARG1_OR_ARG2 = opBinding -> {
+  public static final SqlReturnTypeInference IF_FUNC_RETURN_TYPE = opBinding -> {
     Preconditions.checkState(opBinding.getOperandCount() == 3);
-    if (!opBinding.isOperandNull(1, false)) {
-      return opBinding.getOperandType(1);
-    } else {
-      return opBinding.getOperandType(2);
+    final RelDataType type1 = opBinding.getOperandType(1);
+    final RelDataType type2 = opBinding.getOperandType(2);
+    if (type1 == type2) {
+      return type1;
     }
+    // If the types mismatch, the non-literal one's type should be picked
+    // i.e. the type of `if(..., 0, bigint_type_field)` should be bigint rather than int
```

Comment:
Can you point me a reference about `if` grammar? Looks like it is used to replace `case` statement but the grammar of `case` seems to be a bit different though ...

Reply:
The grammar can be found [here](https://cwiki.apache.org/confluence/display/hive/languagemanual+udf#LanguageManualUDF-ConditionalFunctions)
But it seems that in Hive, the second and third type don't need to be exactly same, SQL like `select if(true, 1, 1.0);` works well, and there are many mismatching types in our production views.

Reply:
can you elaborate more on why we are pick non-literal type as part of comments?

Reply:
Thanks for reminding, because the non-literal one (which is usually a field of table) is picked by Trino and we need to ensure the type picked by Trino is the same as the type returned by coral.

Also added it in the PR summary.